### PR TITLE
Add administrate dashboard

### PR DESCRIPTION
**Pull request by schmijos**
I'd like to drop the JS dashboard in order to deprecate the old REST API. What's necessary in order to reach that @dbrgn ?
At least I'd like to remove all **writing** access from the APIs (except for the measurement registration (`POST /api/measurements`)).

 (or at least remove the writing parts except for measurement creation).

image

### CODE REVIEW

File: app/controllers/admin/application_controller.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+# you're free to overwrite the RESTful controller actions.
+module Admin
+  class ApplicationController < Administrate::ApplicationController
+    include ActionController::HttpAuthentication::Basic::ControllerMethods
+
+    before_action :authenticate_admin
+
+    def authenticate_admin
+      authenticate_or_request_with_http_basic do |given_name, given_password|
+        ApiConsumer.exists?(public_api_key: given_name, private_api_key: given_password)
```
Comment (schmijos) [inline app/controllers/admin/application_controller.rb:15]: Initially the idea was that `public_api_key` is a consumers public name for read access and `private_api_key` is for writing only. Not sure if this concept still holds.

File: config/application.rb [hunk 2]
```diff
@@ -29,7 +29,7 @@ class Application < Rails::Application
     # Only loads a smaller set of middleware suitable for API only apps.
     # Middleware like session, flash, cookies can be added back manually.
     # Skip views, helpers and assets when generating a new resource.
-    config.api_only = true
+    config.api_only = false
```
Comment (schmijos) [inline config/application.rb:32]: Using Session, Flash, MimeRespond etc. now.

File: app/controllers/admin/application_controller.rb (hunk 1 above)
Comment (dbrgn) [reply app/controllers/admin/application_controller.rb:15]: Yep, this is still being used that way. Right now we have 1 external user who uses a public-only key. Additionally, the web application and the apps must contain a hardcoded key, that one is read-only as well.

I'd like to keep the differentiation between readonly and read/write keys. The admin interface should only be accessible using the read/write (private) API key.

Would it be hard to slap a session based login page in front of the admin interface? Basic auth has its downsides, for example: You cannot clear a basic auth session without closing and re-opening the browser. It's not that important though, we can keep the basic auth login for now. What about using the contact_email field for the username? Or just the string "admin"?

File: app/dashboards/api_consumer_dashboard.rb [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  # a hash that describes the type of each of the model's fields.
+  #
+  # Each different type represents an Administrate::Field object,
+  # which determines how the attribute is displayed
+  # on pages throughout the dashboard.
+  ATTRIBUTE_TYPES = {
+    id: Field::Number,
+    public_api_key: Field::String,
+    private_api_key: Field::String,
+    contact_email: Field::String,
```
Comment (dbrgn) [inline app/dashboards/api_consumer_dashboard.rb:14]: One thing I noticed: The web UI enforces that a contact_email must be set, but the database field is not actually `not null`. Wouldn't it be good if the constraints are consistent in the database and in the API, where possible?

In Django, there are two configuration fields in the models: "null" (for the database) and "blank" (for the validation). For fields that should not be empty, I usually set both to `False`.

**Review by dbrgn — COMMENTED** [review]
This is nice! 🙂

Regarding deprecation of the old API: Is it an added effort to keep it? There are some use cases I have in mind, where it could still be handy. For example, we could create a page which lists all our sponsors, and the number of sensors they are sponsoring. With the REST endpoints, that's easy to do. With the mobile app endpoints, it isn't (because they're optimized for a certain navigation flow).

In contrast to the "mobile endpoints" URLs, the REST API is more generic/neutral, so it may enable use cases that we don't have in mind yet.

File: app/controllers/admin/application_controller.rb (hunk 1 above)
Comment (schmijos) [reply app/controllers/admin/application_controller.rb:15]: Sure, I can do that.

File: app/dashboards/api_consumer_dashboard.rb (hunk 3 above)
Comment (schmijos) [reply app/dashboards/api_consumer_dashboard.rb:14]: This is inconsistent, correct. But the inconsistency is between Rails model and database, not here.

I made a PR: https://github.com/gfroerli/api/pull/91

**Pull request merged**